nftables-snippets.nix

```diff
@@ -0,0 +1,159 @@
+flakes: {
+  config,
+  lib,
+  ...
+}:
+with lib; let
+  cfg = config.networking.nftables.firewall.snippets;
+  localZoneName = config.networking.nftables.firewall.localZoneName;
+
+  ruleTypes = ["rule" "policy"];
+in {
+  imports = [
+    (import ./nftables-zoned.nix flakes)
+  ];
+
+  options.networking.nftables.firewall.snippets = {
+    nnf-common = {
+      enable = mkEnableOption (mdDoc "the nnf-common firewall snippet");
+    };
+
+    nnf-conntrack = {
+      enable = mkEnableOption (mdDoc "the nnf-conntrack firewall snippet");
+    };
+
+    nnf-drop = {
+      enable = mkEnableOption (mdDoc "the nnf-drop firewall snippet");
+    };
+
+    nnf-loopback = {
+      enable = mkEnableOption (mdDoc "the nnf-loopback firewall snippet");
+    };
+
+    nnf-dhcpv6 = {
+      enable = mkEnableOption (mdDoc "the nnf-dhcpv6 firewall snippet");
+    };
+
+    nnf-icmp = {
+      enable = mkEnableOption (mdDoc "the nnf-icmp firewall snippet");
+      ipv6Types = mkOption {
+        type = types.listOf types.str;
+        default = ["echo-request" "nd-router-advert" "nd-neighbor-solicit" "nd-neighbor-advert"];
+        description = mdDoc ''
+          List of allowed ICMPv6 types.
+        '';
+      };
+      ipv4Types = mkOption {
+        type = types.listOf types.str;
+        default = ["echo-request" "router-advertisement"];
+        description = mdDoc ''
+          List of allowed ICMP types.
+        '';
+      };
+    };
+
+    nnf-ssh = {
+      enable = mkEnableOption (mdDoc "the nnf-ssh firewall snippet");
+    };
+
+    nnf-nixos-firewall = {
+      enable = mkEnableOption (mdDoc "the nnf-nixos-firewall firewall snippet");
+    };
+  };
+
+  config = mkMerge [
+    (mkIf cfg.nnf-common.enable {
+      networking.nftables.firewall.enable = true;
+      networking.nftables.firewall.snippets = mkDefault {
```

Comment:
This `mkDefault` should probably be distributed to each of the enable options below, otherwise setting any `snippet.<xyz>.enable` while also using nnf-common will override this whole block, disabling the others.

Reply:
I was under the impression, that overrides (which are used by `mkDefault`) were pushed down to the leaf-options by `pushDownProperties` (https://github.com/NixOS/nixpkgs/blob/master/lib/modules.nix#L874). Did you actually observe the behavior you are describing?

Reply:
No sorry, you are right. I was under the impression that `networking.nftables.firewall.snippets` is a option of submodules (which it isn't), in which case the properties would not be pushed down.